### DISCUSSION TITLE
WithFileUploads on Google App Engine Standard

### DISCUSSION THREAD

**Discussion by protoapostoli** [livewire/livewire#5719] [1. Help] [answered]
So I have everything working fine on local development using artisan serve, but when I deploy, because of Google App Engine Standard environment constraints, I'm unable to upload files.

As it's starting to upload the file, it throws this error:
```file_put_contents(/tmp/framework/cache/facade-c2f53e2d655810ac111b48d42b34b209861bb985.php): Failed to open stream: No such file or directory```

Which I traced back to this function:
```
    public function startUpload($name, $fileInfo, $isMultiple)

    {

        if (FileUploadConfiguration::isUsingS3()) {

            throw_if($isMultiple, S3DoesntSupportMultipleFileUploads::class);


            $file = UploadedFile::fake()->create($fileInfo[0]['name'], $fileInfo[0]['size'] / 1024, $fileInfo[0]['type']);


            $this->emit('upload:generatedSignedUrlForS3', $name, GenerateSignedUploadUrl::forS3($file))->self();


            return;

        }

        $this->emit('upload:generatedSignedUrl', $name, GenerateSignedUploadUrl::forLocal())->self();

    }
```

As far as I can tell, it's trying to create a new facade with GenerateSignedUploadUrl::forS3($file). The problem is, with Google App Engine, it only allows things to be created in a specific way in the tmp folder (which I have set up).

So my options are to either forsake Livewire file upload and build some work around, or manually change how livewire deals with this. Having it access the GenerateSignedUrl directly instead of through the facade, following the solution here: https://stackoverflow.com/questions/75519174/laravel-livewire-class-facades-livewire-generatesigneduploadurl-not-found

So my question:

What's the best route for managing this? Do I forsake Livewire's amazing file upload and hand make something dumb and unintuitive, or do I change Livewire's code, and force it to work?

Or is there something I'm missing?

**Comment by protoapostoli** [comment, marked as answer]
My solution to get this working with Google App Engine in a Standard Environment was to use composer to override the specific files I needed to modify in the Livewire vendor directory.

Following some hints from this q&a (https://stackoverflow.com/questions/75519174/laravel-livewire-class-facades-livewire-generatesigneduploadurl-not-found), I was able figure out how to get file uploads to work.

- First, I created a directory: app/Overrides
  - Within this dir, I copied over the files I needed to operate on:
    - vendor/livewire/livewire/src/WithFileUploads.php
    - vendor/livewire/livewire/src/GenerateSignedUploadUrl.php
- In WithFileUploads.php I made the following changes:
  - Change this line
    - use Facades\Livewire\GenerateSignedUploadUrl;
  - To This
    - use Livewire\GenerateSignedUploadUrl; 
- In GenerateSignedUploadUrl.php I made the following changes:
  - Make all functions static.
  - On line 67, change $this-> to self::
 
Now we have almost identical copies of the files from vendor/livewire, with modifications that will allow it to work on GAE Standard.
But we're not done yet, we need to somehow tell the system to load our new files instead of the ones from vendor/livewire.

In order to do this, we need two additions to our composer.json file.
Within the main autoload, add these two lines:
```
"exclude-from-classmap": ["vendor/livewire/livewire/src/WithFileUploads.php", "vendor/livewire/livewire/src/GenerateSignedUploadUrl.php"],
"files": ["app/Overrides/WithFileUploads.php", "app/Overrides/GenerateSignedUploadUrl.php"]
```

And that's it. This was my solution. If there are better alternatives to overriding, I'm very much open to ideas. In fact, I'd love to hear about other solutions if there are any. This just seemed the quickest and easiest to make everything work.

**Reply by halestar** [reply]
Hi,
Thank you so much for this answer! It allowed me to solve my issue as well. I would like to add that the structure changed in Livewire 3, so if you're like me and trying to do a file upload in Livewire 3 over GAE, take note of the following:

* The 2 files to copy have changed!
    * vendor/livewire/livewire/src/Features/SupportFileUploads/WithFileUploads.php
    * vendor/livewire/livewire/src/Features/SupportFileUploads/GenerateSignedUploadUrl.php
* Do the same as above for the WithFileUploads.php file
* In GenerateSignedUploadUrl.php the following changes need to be made:
    * Change the namespace to `namespace Livewire;`
    * At the top also add: `use Livewire\Features\SupportFileUploads\FileUploadConfiguration;` and `use Livewire\Features\SupportFileUploads\TemporaryUploadedFile;`
    * Make all functions static.
    * On line **54**, change `$this->` to `self::`

Finally, the composer file should be:
```
"exclude-from-classmap": ["vendor/livewire/livewire/src/Features/SupportFileUploads/WithFileUploads.php", "vendor/livewire/livewire/src/Features/SupportFileUploads/GenerateSignedUploadUrl.php"],
"files": ["app/Overrides/WithFileUploads.php", "app/Overrides/GenerateSignedUploadUrl.php"]
```

Hope this helps other as much as your answer helped me!
